---
title: "RiTMIC: RegulatIon of Tumor MIcroenvironment Composition"
subtitle: "Part 1: Deconvolution of the bulk samples"
author: "Clementine Decamps, Magali Richard"
contact: 
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
vignette: >
  %\VignetteIndexEntry{deconvolution}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=TRUE, results="verbatim", dpi=75)
layout(1, respect=TRUE)
require(medepir)
library(ggplot2)
```

`RiTMIC` (**R**egulat**I**on of **T**umor **MI**croenvironment **C**omposition) is an open-access R package to study the link between the gene expression in a tumor and the microenvironment composition.  

![](images/pipeline_ritmic.png)

The pipeline is in 4 steps:  

1. **Deconvolution** of the bulk samples to obtain the cell type composition.
2. **Extraction** of the cancer expression.
3. **Differential** analysis of the gene expression in the cancer component.
4. **Analyze** of the link between the gene expression in the cancer and the tumor composition.

---

This vignette allows to apply the **step 1**, the deconvolution step to obtain the cell-type composition. This step is made through the package `medepir`.

# Load data

D_met and D_rna are the bulk samples, respectively for the methylome and the transcriptome of the samples. Here, we used 30 samples simulated with the package Ritmic_simu, with 10,000 probes for the methylome and 1,000 genes for the transcriptome.

```{r, label="load_data"}
D_met = readRDS("data_example/D_met.rds")
D_rna = readRDS("data_example/D_rna.rds")

dim(D_met)
dim(D_rna)
```

# Compute the proportion matrix (A) 

To obtain the matrix of cell types proportions, we used the methylome data. 

First, we pre-treat the data. When you have clinical data, you can use `medepir::CF_detection` to remove the probes correlated with confounding factors. Then, we use `medepir::feature_selection` to reduce the sample size.

With `medepir::plot_k` you can plot a PCA to choose the number of components to deconvolute. Here, we find k = 4.

Then, we compute the cell type proportion matrix with `medepir::Edec`.

```{r, label="Deconv meth"}
D_FS = medepir::feature_selection(D = D_met, number = 2000)
medepir::plot_k(D_FS)
results_edec = medepir::Edec(D_FS, nbcell = 4, infloci = rownames(D_FS))

if(!file.exists("data_example/res_edec.rds")){
  saveRDS(results_edec, file = "data_example/res_edec.rds")
}
```

# Compute the cell type profiles (T) 

To obtain the cell type profiles in RNAseq, we used the A matrix compute on the methylome and the bulk sample of RNAseq. It's the stage 2 function of the `EDec` method which allows this.

```{r, label="Deconv rnaseq"}
prop_cell_type = t(results_edec$A)
res_edec2 = EDec::run_edec_stage_2(D_rna, prop_cell_type)
T_res = res_edec2$means
colnames(T_res) = 1:4

if(!file.exists("data_example/res_edec2.rds")){
  saveRDS(T_res, file = "data_example/res_edec2.rds")
}
```

# Visualization 

## A matrix

To visualize the A matrix, we can use a stacked barplot to see the variation of cell type proportions across the samples.

```{r, label="Visu A"}
df = data.frame(data = c(results_edec$A),
                    type = as.factor(rep(1:4)),
                    p = rep(1:ncol(results_edec$A), each = 4))
    

ggplot(df, aes(x = p, y = data, fill = type, color = type)) +
      geom_bar(stat="identity") +
      theme_minimal() + labs(title = "A matrix")
```

## T matrix

To visualize the T matrix, we can for example use a heatmap, to see the correlation between the deconvolued cell types.

```{r, label="Visu T"}
ComplexHeatmap::Heatmap(cor(T_res))
```

