---
title: "RiTMIC: RegulatIon of Tumor MIcroenvironment Composition"
subtitle: "Part 1 : Simulation of complex gene expression matrix"
author: "Magali Richard, Clementine Decamps, Fabien Quinquis"
contact: 
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
vignette: >
  %\VignetteIndexEntry{simulation}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=TRUE, results="verbatim", dpi=75)
layout(1, respect=TRUE)

if(!dir.exists("tmp_ritmic")){
  dir.create("tmp_ritmic")
}
```

# Introduction

`RiTMIC` (**R**egulat**I**on of **T**umor **MI**croenvironment **C**omposition) is an open-access R package to study the link between the gene expression in a tumor and the microenvironment composition. 

**To do** : describe the pipeline's steps : 1. simulation of complex D matrix / 2. deconvolution with medepir / 3. identification and extraction of the cancer type / 4. Penda / 5. res analysis.

`RiTMIC` Part 1: **Simulation of complex gene expression matrix**

#Generation of dataset

## Step 1: A matrix

The function `simu_A` simulates the matrix of proportion of cell lines distribution per tumor. 

```{r, echo = F, eval = T}
cell_type_proportions = c(1.5, 4.5, 1, 3)
tumor_number = 30
if(!file.exists("tmp_ritmic/A_matrix.rds")){
  A_matrix = Ritmic::simu_A(tumor_number, cell_type_proportions)
  saveRDS(A_matrix, file="tmp_ritmic/A_matrix.rds")
}
A_matrix = readRDS("tmp_ritmic/A_matrix.rds")
```

```{r, label="RiTMIC::simu_A", echo = T, eval = F}
cell_type_proportions = c(1.5, 4.5, 1, 3)
tumor_number = 30
A_matrix = Ritmic::simu_A(tumor_number, cell_type_proportions)
```

## Step 2: T Matrix 

**To do** : explain that we simulated 3 reference cell type profil for RNAseq, 4 for DNAm, and for the RNAseq we simulated one profile/tumor.

###T matrix for RNAseq

**To do** : explain that profiles of 3 cell types for RNAseq are simulated randomly (maybe real dataset in ritmic later)

```{r, label="RiTMIC::simu_T_rna"}
T_rna = matrix(data = sample(seq(0, 5000, 0.1), size = 1000*3, replace = T), nrow = 1000, ncol = 3)
colnames(T_rna) = c("Epi", "IC", "Fibroblast")
rownames(T_rna) = rownames(penda::penda_data_case)[1:1000]

if(!file.exists("tmp_ritmic/T_rna.rds")){
  saveRDS(T_rna, file = "tmp_ritmic/T_rna.rds")
}

dim(T_rna)
head(T_rna)
```


###T matrix for DNAm

**To do** : explain that profiles of 4 cell types for DNAm are simulated randomly (maybe real dataset in ritmic later)

```{r, label="RiTMIC::simu_T_met"}
T_met = matrix(data = sample(seq(0, 1, 0.0001), size = 10000*4, replace = T), nrow = 10000, ncol = 4)
colnames(T_met) = c("Epi", "IC", "Fibroblast", "Tumor")
rownames(T_met) = 1:10000

if(!file.exists("tmp_ritmic/T_met.rds")){
  saveRDS(T_met, file = "tmp_ritmic/T_met.rds")
}

dim(T_met)
head(T_met)
```

###T cancer for RNA_seq

**To do** : Explain that : New simulated RNAseq samples can be obtained from a RNAseq dataset with `simu_T_cancer`: for each gene, novel gene expressions will be assigned by a random choice inside the bimodal distribution of the actual gene expression   

```{r, label="RiTMIC::simu_T_cancer"}
tumor_ref =  penda::penda_data_case[1:1000, 1:10]
dim(tumor_ref)

T_tum <- Ritmic::simu_T_cancer(tumor_ref, tumor_number)
if(!file.exists("tmp_ritmic/T_tum.rds")){
  saveRDS(T_tum, file = "tmp_ritmic/T_tum.rds")
}

dim(T_tum)
head(T_tum)[, 1:5]
```

## Step 3: Deregulation of gene expression according to ME proportion in tumors

**To do** : Two methods to disturb the tumor micro environments are available by a random modification of the gene expressions  :  `corr_prop_f` and `corr_prop_t`.
Methods differ by the calcul of coefficients:

- `corr_prop_f`: newGeneExpression = geneExpression \* (1 + dereg_coeff \* cellTypeProportion))

- `corr_prop_t`: newGeneExpression = geneExpression \* dereg_coeff if thres > cellproportion

Explain here we used factor 


```{r, label="RiTMIC::corr_prop_f"}
T_tum_dereg = Ritmic::corr_prop_f(T_cancer = T_tum,A_ME = A_matrix[3, ], G = 50, dereg_coeff = 10)

T_tum_dereg$g_dereg

if(!file.exists("tmp_ritmic/T_tum_dereg.rds")){
  saveRDS(T_tum_dereg, file = "tmp_ritmic/T_tum_dereg.rds")
}
```

## Step 4 : compute D matrices

**To do** : explain D matrices

###DNAm

**To do** : comment

```{r, label="RiTMIC::simu_D_met"}
D_met = T_met %*% A_matrix

if(!file.exists("tmp_ritmic/D_met.rds")){
  saveRDS(D_met, file = "tmp_ritmic/D_met.rds")
}

dim(D_met)
head(D_met)[, 1:5]
```

###RNAseq

**To do** : comment

```{r, label="RiTMIC::simu_D"}
D_rna <- Ritmic::simu_D(A = A_matrix, T = T_rna, T_cancer = T_tum_dereg, noise = F)

if(!file.exists("tmp_ritmic/D_rna.rds")){
  saveRDS(D_rna, file = "tmp_ritmic/D_rna.rds")
}

dim(D_rna)
head(D_rna)[, 1:5]
```


# Session Information

```{r, results="verbatim"}
sessionInfo()
```