---
title: "RiTMIC: RegulatIon of Tumor MIcroenvironment Composition"
subtitle: "Part 3: Differential analysis of gene expression with `penda`"
author: "Clementine Decamps, Magali Richard"
contact: 
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
vignette: >
  %\VignetteIndexEntry{penda_analysis}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=TRUE, results="verbatim", dpi=75)
layout(1, respect=TRUE)
```

`RiTMIC` (**R**egulat**I**on of **T**umor **MI**croenvironment **C**omposition) is an open-access R package to study the link between the gene expression in a tumor and the microenvironment composition.  

![](images/pipeline_ritmic.png)

The pipeline is in 4 steps:  

1. **Deconvolution** of the bulk samples to obtain the cell type composition.
2. **Extraction** of the cancer expression.
3. **Differential** analysis of the gene expression in the cancer component.
4. **Analyze** of the link between the gene expression in the cancer and the tumor composition.

---

This vignette allows to apply the **step 3**, the step of differential gene expression analysis.   

For this step, we used `penda` (**PE**rso**N**alized **D**ifferential **A**nalysis ), an open-access R package that detects gene deregulation in individual samples compared to a set of reference, control samples. 

# Dataset

$D_{cancer}$ is the matrix with the gene expression of cancer cells in each sample, obtained in the vignette 2_identification_extraction_tumor_expression.Rmd.

```{r, label="load_data"}
D_cancer = readRDS("data_example/D_cancer_pretreat.rds")
```

# Penda method

`penda` performs a 3-steps analysis: 
1. Data filtering
2. Relative gene ordering
3. Differential expression testing  

To use Penda, we need a control sample. Here, we use the example dataset penda::penda_data_ctrl, but for a real analysis you can use, for example, normal cell lines.

## Data filtering

This step arranges data in penda format.
```{r echo=FALSE, label="data filtering"}
penda_data_case = D_cancer
penda_data_ctrl = penda::penda_data_ctrl[1:1000, 1:10]

Penda_dataset = penda::make_dataset(penda_data_ctrl, penda_data_case, detectlowvalue = F)
data_ctrl = Penda_dataset$data_ctrl
data_case = Penda_dataset$data_case
```

## Relative gene ordering

This step make the reference rank of genes in controls data.

```{r, label= "penda::compute_lower_and_higher_lists"}
s_max = 30
L_H_list = penda::compute_lower_and_higher_lists(data_ctrl, threshold = 0.99, s_max = s_max)
L = L_H_list$L
H = L_H_list$H
```

## Differential expression testing

Finally, this step is the test of deregulation. You can fit the penda parameters by making simulations, see penda package for more information.

```{r, label="penda::penda_test", results="hide"}
threshold = 0.4
iterations =  20

penda_res = penda::penda_test(samples = data_case, 
                              controls = data_ctrl,
                              iterations =  iterations, 
                              L_H_list =  L_H_list, 
                              threshold = threshold)

if(!file.exists("data_example/penda_res.rds")){
  saveRDS(penda_res, file = "data_example/penda_res.rds")
}

prop.table(table(penda_res$down_genes))
prop.table(table(penda_res$up_genes))

```


