---
title: "RiTMIC: RegulatIon of Tumor MIcroenvironment Composition"
subtitle: "Step 1 : Increase your RNAseq datasets with a new simulated adapted dataset derivated from yours"
author: "Magali Richard, Clementine Decamps, Florent Chuffart, Fabien Quinquis, Daniel Jost"
contact: 
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
vignette: >
  %\VignetteIndexEntry{simulation}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=TRUE, results="verbatim", dpi=75)
layout(1, respect=TRUE)
```

# Introduction

`RiTMIC` (**R**egulat**I**on of **T**umor **MI**croenvironment **C**omposition) is an open-access R package that simulates the necessary tools to perform `PenDA` analysis, perform analysis with `PenDA` and apply visualization tools to the `PenDA` output. 

`RiTMIC` Step 1: **Inputs required to `penda` are simulated with `simulation` functions.**

# Generation of input files for penda analysis

`PenDA` analysis is an open-access R package that detects gene deregulation in individual samples compared to a set of reference, control samples. 

One matrix of gene expression per sample is required to perform. This matrix can be directly used in `PenDA`.

`simulation` has been developed to simulate an increased dataset that fits with your RNAseq gene distribution dataset. This functionality requires the different cell proportions distributions.

## Explanation of dataset

`data_ctrl` is a data matrix containing the normalized counts of the control sample. 
The names of the matrix correspond to the gene_symbol.

`data_case` is a data matrix containing the normalized counts of each tumor sample. 
The rownames of the matrix correspond to the gene_symbol, the colnames indicate the sample ID.

## Increase your dataset with a randomized gene expression distribution

`RiTMIC::simu_T_cancer` 
