---
title: "RiTMIC: RegulatIon of Tumor MIcroenvironment Composition"
subtitle: "Part 3 : Statistical analysis of penda outputs"
author: "Magali Richard, Clementine Decamps, Florent Chuffart, Fabien Quinquis, Daniel Jost"
contact: 
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
vignette: >
  %\VignetteIndexEntry{statistics}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=TRUE, results="verbatim", dpi=75)
layout(1, respect=TRUE)
```


# Introduction

`RiTMIC` (**R**egulat**I**on of **T**umor **MI**croenvironment **C**omposition) is an open-access R package that simulates the necessary tools to perform `PenDA` analysis, perform analysis with `PenDA` and apply visualization tools to the `PenDA` output. 

`RiTMIC` Part 3: **Statistical analysis of PenDA outputs**

# Statistical analysis 

`statistics` has been developed to show the model improvement  thanks to the __PenDA__ analysis. 

The ROC curve is plotted at the end of the pipeline to observe __different distance methods(ks, student, kanto) versus the standard correlation comparison__.

`PenDA` analysis is an open-access R package that detects gene deregulation in individual samples by relative gene ordering and differential expression testing. This analysis will be performed on step 2 of the pipeline.

# Understanding results 

If curves of distance methods has a higher True Positive Rate (TPR) combined a lower False Positive Rate(FPR) than correlation curve, __PenDA__ improves the model.  

# Presentation of the input dataset

Three data input are required to perform the statistical analysis :

- penda::penda_test_1ctrl output : the down and up matrices of dimension: n*g 
  - g: gene names 
  - n: the deregulation status of the cancer compared to the control
    - -1: Down-regulated gene 
    - 0: Same expression
    - +1: Up-regulated gene 

- Matrix A of cell lines proportion distribution simulated in `step_1_simulation`
  - proportion: k*n
  - n: patient 
  - k: proportion of the different cell lines
  
- Matrix T of reference cell lines simulated for the example in `step_1_simulation`
  - proportion: ge x nc
  - ge: gene expression
  - nc: number of cell lines 
  
## Step 1: Loading required dataset and apply RiTMIC::pre_treat function

```{r, label="loading_data_set_and_pre_treat_function"}
# Penda::penda_test_1_ctrl output 
test_matrix <- readRDS("~/Datas/projects/RiTMIC/inst/extdata/penda_res.RDS")
dim(test_matrix$down_genes)
dim(test_matrix$up_genes)

# Matrix_A of cell line proportions distribution per sample
matrix_A <- readRDS("~/Datas/projects/RiTMIC/inst/extdata/matrix_A.RDS")
dim(matrix_A)

# Matrix T control 
simu_D_output <- readRDS("~/Datas/projects/RiTMIC/inst/extdata/matrix_D.RDS")
pre_treat_res <- RiTMIC::pre_treat(penda_res = test_matrix, patientNumber=5)

```

```{r,label="RiTMIC::calc_dist", message=FALSE}
distance <- RiTMIC ::calc_dist(pre_treat_output = pre_treat_res, matrix_A = matrix_A)
calc_corr <- RiTMIC::calc_corr(matrix_D = simu_D_output, matrix_A = matrix_A)
```
## Did PenDA had an impact to the simulated results ? 

To build this vignette, we needed to do some approximation: the controle matrix is not related to a real T matrix dataset. 
So, a greater correlation score than the distance scores simulated with kanto, ks, student do not reveal a true result. 

To have a reliable result, please run the pipeline with real dataset. 
```{r,label="RiTMIC::plot_res"}
RiTMIC::plot_res(calc_corr_output = calc_corr, calc_dist_output = distance, graph_title = "ROC curves to test the model improvement induce by PenDA")
```

# Session Information

```{r, results="verbatim"}
sessionInfo()
```